Modules/_datetimemodule.c

```diff
@@ -6386,18 +6386,35 @@ PyInit__datetime(void)
     if (m == NULL)
         return NULL;
 
-    if (PyType_Ready(&PyDateTime_DateType) < 0)
+    if (PyModule_AddType(m, &PyDateTime_DateType) < 0) {
+        Py_DECREF(m);
         return NULL;
-    if (PyType_Ready(&PyDateTime_DateTimeType) < 0)
+    }
+
+    if (PyModule_AddType(m, &PyDateTime_DateTimeType) < 0) {
+        Py_DECREF(m);
         return NULL;
-    if (PyType_Ready(&PyDateTime_DeltaType) < 0)
+    }
+
+    if (PyModule_AddType(m, &PyDateTime_TimeType) < 0) {
+        Py_DECREF(m);
         return NULL;
-    if (PyType_Ready(&PyDateTime_TimeType) < 0)
+    }
+
+    if (PyModule_AddType(m, &PyDateTime_DeltaType) < 0) {
+        Py_DECREF(m);
         return NULL;
-    if (PyType_Ready(&PyDateTime_TZInfoType) < 0)
+    }
+
+    if (PyModule_AddType(m, &PyDateTime_TZInfoType) < 0) {
+        Py_DECREF(m);
         return NULL;
-    if (PyType_Ready(&PyDateTime_TimeZoneType) < 0)
+    }
+
+    if (PyModule_AddType(m, &PyDateTime_TimeZoneType) < 0) {
+        Py_DECREF(m);
```

Comment:
I'm not entirely clear on what happens if you DECREF `m` here - will it also automatically `DECREF` all the stuff that's been added to the module by this point?

I think it's not a major issue because this will only happen if you fail to initialize the `datetime` module, and if that happens I imagine resource leaks are the least of your problems (looks to me like there are a lot of "tolerable refleaks" like that elsewhere in the initialization code), but it would be nice to get it right if we can, I guess.

Another possibility here that might be more DRY might be to use a for loop?

```c
static PyTypeObject *types = {
    &PyDateTime_DateType,
    &PyDateTime_DateTimeType,
    ...
}

for (size_t i = 0; i < sizeof(types); ++i) {
    if (PyModule_AddType(m, types[i]) < 0) {
        Py_DECREF(m);
        return NULL;
    }
}
```

Reply:
FYI, we can use `Py_ARRAY_LENGTH` for this situation :)
I learned it from Victor :)
 https://github.com/python/cpython/pull/19088#discussion_r396017489

Reply:
I see that Py_DECREF(m) has been removed. That's a bug. But I suggest to open a new PR/issue to enhance PyInit__datetime() error handling. I'm fine with only using "return NULL" to "be consistent" with other error handling paths of this function for now :-)

@corona10: Do you want to attempt to either port _datetime to multiphase initialization, or at least enhance its error handling, in a separated PR/issue?

Reply:
To clarify my original concern: I liked the `Py_DECREF(m)`, what I was concerned about is that each successful call to `PyModule_AddType(m, type)` increases the reference count of `type`, and I wasn't sure if `m` *steals* a reference to `type`. 

Looking at [`PyModule_AddObject`](https://docs.python.org/3/c-api/module.html#c.PyModule_AddObject), I guess the answer is that yes, it *does* steal a reference, so `Py_DECREF(m)` is enough.

I think for now let's keep it with the for loop and the `Py_DECREF(m)` and we can do a separate change to clean up the reference leaks later (maybe just add a `goto error` that decrefs the module and returns `NULL`.)

@vstinner Looks like someone already has a PR open to add multiphase initialization to `datetimemodule.c`: #19122.

Reply:
#19122 can be conflicted with this change

Reply:
> But I suggest to open a new PR/issue to enhance PyInit__datetime() error handling

I like this idea. The current usage is not chaged on this PR.